src/data_race.rs

```diff
@@ -544,31 +544,42 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: MiriEvalContextExt<'mir, 'tcx> {
 
     /// Perform an atomic compare and exchange at a given memory location.
     /// On success an atomic RMW operation is performed and on failure
-    /// only an atomic read occurs.
+    /// only an atomic read occurs. If `can_fail_spuriously` is true,
+    /// then we treat it as a "compare_exchange_weak" operation, and
+    /// some portion of the time fail even when the values are actually
+    /// identical.
     fn atomic_compare_exchange_scalar(
         &mut self,
         place: MPlaceTy<'tcx, Tag>,
         expect_old: ImmTy<'tcx, Tag>,
         new: ScalarMaybeUninit<Tag>,
         success: AtomicRwOp,
         fail: AtomicReadOp,
+        can_fail_spuriously: bool,
     ) -> InterpResult<'tcx, Immediate<Tag>> {
+        use rand::Rng as _;
         let this = self.eval_context_mut();
 
         // Failure ordering cannot be stronger than success ordering, therefore first attempt
         // to read with the failure ordering and if successful then try again with the success
         // read ordering and write in the success case.
         // Read as immediate for the sake of `binary_op()`
         let old = this.allow_data_races_mut(|this| this.read_immediate(place.into()))?;
-
         // `binary_op` will bail if either of them is not a scalar.
         let eq = this.overflowing_binary_op(mir::BinOp::Eq, old, expect_old)?.0;
-        let res = Immediate::ScalarPair(old.to_scalar_or_uninit(), eq.into());
+        // If the operation would succeed, but is "weak", fail 50% of the time.
+        // FIXME: this is quite arbitrary.
```

Comment:
That is a pretty cool idea. I think we should have a new flag for these kind of operations (for tuning and enabling/disabling), since I don't think it fits any of the existing disable flags that we have.

Reply:
Yeah it is quite cool indeed. :) However, I wonder if a flag is overkill? When would one want to disable this?

I'm actually wondering if we should use more than 50% chance of failure. After all, we want a reasonably high chance of the program going around the loop at least once.

Reply:
well... ideally we'd guarantee that it is getting run at least two or three times. So we could make the first few calls always fail. This can't be done reliably though, since we have no way of knowing what is a repetition of the previous attempt, the user may use arbitrary repetition schemes.

So the only flag I think we need is for tuning the failure chance.

Reply:
> When would one want to disable this?

This was my stance too sort of. In practice if I'm running in miri I want a higher chance of seeing a bug, especially since I have to [crank down test iterations under miri when testing concurrent code](https://github.com/thomcc/cobb/blob/main/examples/stack.rs#L69).

That said, I guess someone might want to disable this if a library they use has a bug here or something, and they don't care because x86. I'm not particularly sympathetic to this case, however.

> After all, we want a reasonably high chance of the program going around the loop at least once.

Yes, I agree here too. I'm open to changing it, but don't have strong feelings as to what to set it to.

Reply:
> So the only flag I think we need is for tuning the failure chance.

Given that we cannot find an obviously right value, this makes sense. I propose something like 0.8 as the default... does anyone remember enough statistics to tell what the expected number of loop iterations is with that value?^^

Reply:
No, but I set it at 100 as a "this indicates something terrible has happened" arbitrary constant. It shouldn't take any extra time during tests anyway, since it will only do those iterations on failure, and in practice even under miri I think it won't take that long.

Reply:
Ah, I'll add a flag for this tomorrow maybe. (I have to figure out how to add a flag)

Reply:
> expected number of loop iterations is with that value?

If each iteration breaks iteration with the probability of t, then the expected number of iterations it `1/t`.

Reply:
> If each iteration breaks iteration with the probability of t, then the expected number of iterations is 1/t.

Thanks! I thought this might be the case but failed to cook up the right search term to confirm...

Reply:
@thomcc

> No, but I set it at 100

(No idea what that comment is responding to.)

> (I have to figure out how to add a flag)

You need to add a new field to

https://github.com/rust-lang/miri/blob/de0800e83b4e15cf3c6aa8f15f8328e86a95d955/src/eval.rs#L26

and to

https://github.com/rust-lang/miri/blob/de0800e83b4e15cf3c6aa8f15f8328e86a95d955/src/machine.rs#L120

and then extend our hacky argument parsing at

https://github.com/rust-lang/miri/blob/de0800e83b4e15cf3c6aa8f15f8328e86a95d955/src/bin/miri.rs#L191

Reply:
> (No idea what that comment is responding to.)

I bumped the iterations in the failure test to 100: https://github.com/rust-lang/miri/pull/1686/files#diff-8fbc7bb0777020b759f156d82407d3f6a583c8f3c5f53111f652a5d548e160edR92

Reply:
I took it as a float between 0.0 and 1.0. A value of 0.0 should never fail, and 1.0 should always fail.